core/src/client/async_client/mod.rs

```diff
@@ -232,11 +231,15 @@ impl Client {
 		*err_lock = next_state;
 		err
 	}
-}
 
-impl Drop for Client {
-	fn drop(&mut self) {
-		self.to_back.close_channel();
+	/// Completes when the client is disconnected or the client's background task encountered an error.
+	/// If the client is already disconnected, the future produced by this method will complete immediately.
+	///
+	/// # Cancel safety
```

Comment:
this is really great, we should adopt this idiom IMO to document cancel safety.

most things here are cancel-safe but really great to be explicit.

Reply:
Agreed! It's so useful to know :)